### DISCUSSION TITLE
Jamulus Server - Direct Input Tracks to [ASIO] Outputs

### DISCUSSION THREAD

**Discussion by ceccareb** [jamulussoftware/jamulus#1361] [Ideas]
The Jamulus server can record client stereo/mono tracks in a Reaper project.   

Given:   Reaper supports ReaRoute which is a virtual ASIO card.   ReaRoute has many inputs and outputs.

Feature:   Add the ability to the Jamulus Server to direct the incoming client tracks to audio inputs of ReaRoute.   

Purpose:   I can now tell Reaper, in real-time, to ingest the audio inputs from Jamulus Server and direct them to separate Reaper tracks.  I can then tell Reaper how to mix each Reaper Track.   I use Reaper as the mixer.   I forward the mix from Reaper to OBS (online broadcasting software).   OBS knows ReaRoute too.   OBS then inputs the output channels from Reaper, combines those channels with video, then records the video or livestreams the video to FaceBook or YouTube.

Example:   I want to tell Jamulus to output my friend Mike's guitar and vocal tracks to ReaRoute's input channels 1 and 2, the bass player Dave's to ReaRoute input channels 3 and 4, etc.   

I use Reaper as the mixer.   I tell Reaper to put ReaRoute 1 and 2 as the source from the guitar and vocal, 3 and 4 for bass and vocal.    I can pan channel 3 to the middle, 4 to right, adjust volume, even add a VST effect.    This is all real-time using the "Audio Monitor = Yes" in Reaper.

Using Reaper, I mix to a stereo out--another ReaRoute channel.

I tell OBS to use the audio input from the ReaRoute channels that were output from Reaper.

**Comment by gilgongo** [comment]
@ceccareb this may need some discussion so I'll move it over there if that's OK.

**Comment by ceccareb** [comment]
There are 2 shortcomings with Jamulus:

1.   I cannot control left and right volumes from a client.   There is only one fader for both.   In my case, the remote client is a guitar player who plays guitar in the left channel and sings in the right.    But my guitar player must combine his left and right into a single stereo mix else he sounds lopsided at my client.   What I need is this:  I need to be able to separate his voice from his guitar and still make the voice and guitar sound in the center.

2.  The Jamulus client is a personal mixer.   While the mix is good for me, the piano player, it is not good for a livestream for the whole band.   Too much piano.   Therefore:

a.   I need a mixer at the Jamulus Server where an objective sound engineer can adjust the volume and pan for the overall good.   He can take the guitarist's vocal and pan it to center.   He can raise the melody vocal up above the other inputs.
b.  A suggestion:   In place of putting a mixer inside the Jamulus Server, one can add code to the Jamulus Server for the Server to direct the input channels from all the clients to separate ASIO outputs.    Why?
c.  Using ReaRoute plugin, Reaper can receive the Jamulus Server outputs via virtual ReaRoute ASIO driver.    Reaper becomes the real-time mixer.
d. Using Reaper's Auto-Monitor=YES, Reaper, one can send the real-time mix to OBS.   From OBS, one can broadcast to Facebook or YouTube.

Because the Jamulus Server already has the ability to record all the separate client channels into a separate tracks in a Reaper project, I figure it is not a big leap to send the client channels instead to ASIO outputs.
